### DISCUSSION TITLE
Discussione Classi

### DISCUSSION THREAD

**Discussion by gaiadobc** [FrancescaR11/ESERCIZIO_FILESYSTEM#6] [General]
Abbiamo provato a seguire i suoi suggerimenti ed abbiamo caricato il nuovo diagramma delle classi e una nuova versione del file, 'NUOVO_FILESYSTEM_CON_CLASSI' coerente con il nuovo diagramma delle classi.

**Comment by iannellog** [comment]
Ho ripensato a quanto ci siamo detti oggi.
Secondo me potete fare la seguente cosa preservando molto del lavoro fatto.
- Il diagramma delle classi lo lasciate come è
- rinunciate al fatto che checkerai restituisce un oggetto condizione, ma gli fate restituire False/True
- partendo dal file JSON create una lista di oggetti condizione: uno per ogni condizione (così la classe astratta Condition va bene come è, dovete solo implementare bene il metodo create_instance)
- il vostro programma principale diventa:

out_file_list = []
for filename in file_list:
    f = FormatReader.create_instance(filename)
    ok = True
    for c in conditions_list:
        if not c.checker(f): # dentro checker usate i metodi di f per esaminarne il contenuto
            ok = False
            break
    if ok:
        out_file_list.append(f.get_name())

- il problema che una condizione non si può applicare a un file lo risolvete facendo un controllo dentro checker e nel caso la condizione non si possa applicare restituite sempre True (che equivale di fatto a "saltare" la condizione per quel file
- in questo modo vi si semplifica tutto molto ed è molto più generale

Pensateci e mercoledì ne parliamo. Se avete dubbi che non possono aspettare rispondete a questo commento così vi contatto prima.

**Comment by iannellog** [comment]
Un altro suggerimento.
Se la classe OccurrenceChecker ha un attributo "filename", gli oggetti della classe vengono associati in modo fisso a uno specifico file. Non sarebbe più logico che il file sia un parametro del metodo checker (vedi precedente commento) e che invece gli attributi siano la parola e il numero di occorrenze cercato?
Domanda: quanto sopra significa creare un oggetto diverso per ogni coppia (parola, occorrenza)?
In questo modo si creano meno oggetti o più oggetti rispetto alla vostra soluzione che lega l'oggetto OccurrenceChecker a uno specifico file?
Provate ad applicare questa idea ad altre condizioni (vedi file allegato)

[Traccia_secondo_progetto_v2.zip](https://github.com/FrancescaR11/ESERCIZIO_FILESYSTEM/files/5798334/Traccia_secondo_progetto_v2.zip)

**Comment by gaiadobc** [comment]
Nel pomeriggio abbiamo lavorato su quello che ci aveva detto e siamo
arrivate ad una soluzione simile a quella che ci propone ora. La differenza
è che non eseguiamo un for sulla lista di condizioni (poichè cosi crediamo
che la lista dovrebbe a quel punto contenere  l’ordine esatto in cui
eseguire le condizioni, volendo mantenere l’idea della struttura ad
albero). Invece, seguendo la sua idea ,facciamo un ciclo while che si
arresta quando esaurisce le condizioni salvando il file se queste ultime
sono verificate , oppure quando le condizioni non sono soddisfatte. La
condizione sul path è una classe a parte. Nel main la chiamata al metodo
checker  restituisce True/False e rimanda alla classe della condizione
successiva.  Se volesse dargli uno sguardo abbiamo caricato il programma su
github. Intanto daremo seguito a quando indicatoci nell’ultima mail. Grazie
del suo aiuto e del supporto, Buona Serata

Il giorno lun 11 gen 2021 alle 22:06 Giulio Iannello <
notifications@github.com> ha scritto:

**Comment by iannellog** [comment]
Ho riflettuto e con la vostra soluzione c'e' un problema.
Che succede se le condizioni richieste cambiano? 
Il file di ingresso deve poter contenere le condizioni che interessano di volta in volta (scelte ovviamente tra quelle consentite che sono ben definite, ma non è detto che interessino tutte per una determinata ricerca 
e alcune condizioni possono essere presenti più' volte). Quindi le condizioni da applicare non sono rigidamente predeterminate.
Allego degli esempi di file JSON che il vostro programma dovrebbe poter elaborare. Questo dovrebbe aiutarvi a identificare i problemi con la vostra soluzione. 
Rileggete anche i miei commenti di ieri: non so se avete provato a rispondere alle domande.
Pensateci con calma e non preoccupatevi se sembra che la vostra soluzione vada in crisi. Non è così e ne parlammo domani.
Piuttosto un'altra osservazione: continuate a usare variabile globali. Non va bene: devono diventare parametri o della classe o di un metodo (non è lo stesso, sono due situazioni diverse, e va valutata quale è la soluzione più opportuna)

[casi_di_test.zip](https://github.com/FrancescaR11/ESERCIZIO_FILESYSTEM/files/5801788/casi_di_test.zip)

**Comment by iannellog** [comment]
Anticipo che avete fatto un buon lavoro. 
Ci sono diverse cose che possono essere migliorate e che possono servire per farvi fare esperienza, ma il grosso è fatto. 
A domani.
